File: packages/efa_platform_client_ts/src/lib/jwt.ts
/**
 * Decodes the `sub` (user id) claim of an access-token JWT without
 * verifying the signature (display metadata only; the server verifies).
 */
export function decodeJwtSubject(token: string): string | null {
    const parts = token.split(".");
    if (parts.length !== 3) return null;
    try {
        const payload = decodeBase64Url(parts[1]);
        const json: unknown = JSON.parse(payload);
        if (typeof json === "object" && json !== null && "sub" in json) {
            const sub = (json as { sub: unknown }).sub;
            return typeof sub === "string" ? sub : null;
        }
    } catch {
        // Malformed token; treated as undecodable.
    }
    return null;
}

function decodeBase64Url(segment: string): string {
    const base64 = segment.replace(/-/g, "+").replace(/_/g, "/");
    const padded = base64.padEnd(base64.length + ((4 - (base64.length % 4)) % 4), "=");
    const binary = atob(padded);
    const bytes = Uint8Array.from(binary, (ch) => ch.charCodeAt(0));
    return new TextDecoder().decode(bytes);
}

File: packages/efa_platform_client/lib/src/jwt.dart
import "dart:convert";

/// Decodes the `sub` (user id) claim of an access-token JWT without
/// verifying the signature (display metadata only; the server verifies).
String? decodeJwtSubject(String token) {
  final parts = token.split(".");
  if (parts.length != 3) return null;
  try {
    final payload = utf8.decode(base64Url.decode(base64Url.normalize(parts[1])));
    final json = jsonDecode(payload);
    if (json is Map<String, dynamic>) return json["sub"] as String?;
  } on Object {
    // Malformed token; treated as undecodable.
  }
  return null;
}
